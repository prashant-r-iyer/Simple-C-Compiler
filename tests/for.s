	.data
	.comm i, 8
	.text

.globl main
main:
	# Save Frame pointer
	pushq %rbp
	movq %rsp, %rbp
	subq $256, %rsp
	pushq %rbx
	pushq %rbx
	pushq %r10
	pushq %r13
	pushq %r14
	pushq %r15
	pushq %r11
	pushq %r9

	# push 0
	movq $0, %rbx
	movq %rbx, i

for_condition_1:
	movq i, %rbx

	# push 15
	movq $15, %r10
	movq $0, %rax
	cmpq %r10, %rbx
	setl %al
	movq %rax, %rbx
	cmpq $0, %rbx
	je for_end_1
	jmp for_statement_1

for_increment_1:
	movq i, %rbx

	# push 1
	movq $1, %r10

	# +
	addq %r10, %rbx
	movq %rbx, i
	jmp for_condition_1

for_statement_1:
	movq $string0, %rbx
	movq i, %r10
	# func=printf nargs=2

	# Move values from reg stack to reg args
	movq %r10, %rsi
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx
	jmp for_increment_1

for_end_1:
	movq $string1, %rbx
	# func=printf nargs=1

	# Move values from reg stack to reg args
	movq %rbx, %rdi
	movl $0, %eax
	call printf
	movq %rax, %rbx

	# Restore registers
	popq %r9
	popq %r11
	popq %r15
	popq %r14
	popq %r13
	popq %r10
	popq %rbx
	popq %rbx
	addq $256, %rsp

	leave
	ret

string0:
	.string "i=%d\n"

string1:
	.string "OK\n"

