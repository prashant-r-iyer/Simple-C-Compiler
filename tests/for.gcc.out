for.c: In function ‘main’:
for.c:7:17: warning: implicit declaration of function ‘printf’ [-Wimplicit-function-declaration]
    7 |                 printf("i=%d\n", i);
      |                 ^~~~~~
for.c:1:1: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
  +++ |+#include <stdio.h>
    1 | 
for.c:7:17: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
    7 |                 printf("i=%d\n", i);
      |                 ^~~~~~
for.c:7:17: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
for.c:7:28: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘long int’ [-Wformat=]
    7 |                 printf("i=%d\n", i);
      |                           ~^     ~
      |                            |     |
      |                            int   long int
      |                           %ld
for.c:10:9: warning: incompatible implicit declaration of built-in function ‘printf’ [-Wbuiltin-declaration-mismatch]
   10 |         printf("OK\n");
      |         ^~~~~~
for.c:10:9: note: include ‘<stdio.h>’ or provide a declaration of ‘printf’
